synapse/util/ratelimitutils.py

```diff
@@ -222,20 +229,21 @@ def on_both(r: object) -> object:
             # Ensure that we've properly cleaned up.
             self.sleeping_requests.discard(request_id)
             self.ready_request_queue.pop(request_id, None)
-            wait_span_scope.__exit__(None, None, None)
-            wait_timer_cm.__exit__(None, None, None)
             return r
 
-        # Tracing
-        wait_span_scope = start_active_span("ratelimit wait")
-        wait_span_scope.__enter__()
-        # Metrics
-        wait_timer_cm = queue_wait_timer.time()
-        wait_timer_cm.__enter__()
-
         ret_defer.addCallbacks(on_start, on_err)
         ret_defer.addBoth(on_both)
-        return make_deferred_yieldable(ret_defer)
+        return ret_defer
+
+    async def _trace_wait(self, d: "defer.Deferred[None]") -> None:
+        """Waits for a given `Deferred` to resolve, while timing it for metrics.
+
+        Args:
+            d: The `Deferred` to wait for. Must not follow the Synapse logging context
+                rules.
+        """
```

Comment:
finding this slightly hard to wrap my head around. Rather than passing in a deferred, I would pass in the request id and have it call `_on_enter`. And do some renaming of things.